File: 0001-0500/0099-Recover-Binary-Search-Tree/java-0099/src/Solution3.java
/// Source : https://leetcode.com/problems/validate-binary-search-tree/description/
/// Author : liuyubobobo
/// Time   : 2018-05-29

/// Using Morris inOrder traversal
/// Find the mistaked swapped nodes during the traversing
///
/// Time Complexity: O(n)
/// Space Complexity: O(1)
class Solution3 {

    private TreeNode pre;
    private TreeNode a, b;

    public void recoverTree(TreeNode root) {

        TreeNode cur = root;
        while(cur != null){
            if(cur.left == null){
                process(cur);
                cur = cur.right;
            }
            else{
                TreeNode prev = cur.left;
                while(prev.right != null && prev.right != cur)
                    prev = prev.right;

                if(prev.right == null){
                    prev.right = cur;
                    cur = cur.left;
                }
                else{
                    prev.right = null;
                    process(cur);
                    cur = cur.right;
                }
            }
        }

        int t = a.val;
        a.val = b.val;
        b.val = t;
    }

    private void process(TreeNode node){

        if(pre != null && pre.val > node.val){
            if(a == null){
                a = pre;
                b = node;
            }
            else
                b = node;
        }

        pre = node;
    }
}

File: 0001-0500/0101-Symmetric-Tree/py-0101/Solution3.py
# Source : https://leetcode.com/problems/symmetric-tree/
# Author : penpenps
# Time   : 2019-07-09

# Level traversal using two queues to keep left and right sub-tree
# Time Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        q1, q2 = [root], [root]
        while q1 and q2:
            n1, n2 = q1.pop(), q2.pop()
            if not n1 and not n2:
                continue
            if not n1 or not n2:
                return False
            if n1.val != n2.val:
                return False
            q1.append(n1.left)
            q1.append(n1.right)
            q2.append(n2.right)
            q2.append(n2.left)
        return not q1 and not q2